### DISCUSSION TITLE
Treating leading zeros consistently and restricting to SemVer versions only

### DISCUSSION THREAD

**Discussion by adamralph** [adamralph/minver#719] [General]
Currently, leading zeros are removed from the major, minor and patch parts, but they are preserved in pre-release identifiers.

For example, the tag `01.02.03-beta.0004` is parsed as version `1.2.3-beta.0004`. This doesn't feel right. It seems logical to either parse as version `01.02.03-beta.0004` (preserving all leading zeros) or as version `1.2.3-beta.4` (preserving no leading zeros).

But that leads into a wider question: both convention and [SemVer 2.0](https://semver.org/spec/v2.0.0.html) put restrictions on which characters can and can't be in a specific version part. I cannot recall seeing anyone using leading zeros in major, minor, or patch parts, and SemVer disallows it, which generally makes it "OK" that MinVer strips out those leading zeros. So effectively, MinVer is restricting the version to _some_ parts of SemVer (and general convention), and transforming _some_ parts of the version to comply on the consumer's behalf. This also feels off. It should either be completely restrictive, or completely unrestrictive (possibly with an option to switch modes). And if it is restrictive, what better standard to use than SemVer? Also, what does "completely unrestrictive" mean? There would probably still have to be _some_ restrictions, and not just free text.

I'm capturing this in a discussion for now because I'm not sure whether this discussion is going to reveal bugs or features, nor how many of each. There may also be two (or more) separate concerns here but I can't disentangle them yet.

Thanks to @bording for bringing this to my attention.

**Comment by adamralph** [comment]
I _think_ https://github.com/adamralph/minver/pull/720 addresses all of this.

**Comment by teneko** [comment]
Hi @adamralph. I've written a SemVer parser I am about to release soon. It is 100% restrictive and customizable when one of these restrictions are hit. Would you be interested in it?

Here the parser and tests:
https://github.com/Teneko-NET-Tools/Teneko.VersionableMessages/blob/main/src/Teneko.VersionableMessages.SemVer.Parser/SemanticVersionParser.cs
https://github.com/Teneko-NET-Tools/Teneko.VersionableMessages/blob/main/test/Teneko.VersionableMessages.SemVer.Parser.Test/SemanticVersionParserTest.cs

You can easily use like I did here in my version class: https://github.com/Teneko-NET-Tools/Teneko.VersionableMessages/blob/main/src/Teneko.VersionableMessages.SemVer/SemanticVersion.cs

**Reply by teneko** [reply]
I was really surprised I could not find a mature SemVer parser. And all current SemanticVersion provider I am aware of are lacking strict parsing or have some loopholes. What is your impression to this?

**Reply by adamralph** [reply]
> I've written a SemVer parser I am about to release soon. It is 100% restrictive and customizable when one of these restrictions are hit. Would you be interested in it?

@teneko thanks for letting me know but I think it's probably better to stick to using [the `SemanticVersion` type in the SDK](https://github.com/NuGet/NuGet.Client/tree/dev/src/NuGet.Core/NuGet.Versioning), as I did in https://github.com/adamralph/minver/pull/720.

> I was really surprised I could not find a mature SemVer parser. And all current SemanticVersion provider I am aware of are lacking strict parsing or have some loopholes. What is your impression to this?

AFAIK the `SemanticVersion` type in the SDK matches the spec with only one exception—it does a [case-insensitive comparison of pre-release identifiers](https://github.com/NuGet/NuGet.Client/blob/e70d3c6167fc1c878acee98050784726152e4a31/src/NuGet.Core/NuGet.Versioning/VersionComparer.cs#L330-L331). The [proposed change to the spec](https://github.com/semver/semver/pull/276) was rejected, but NuGet went ahead and implemented it anyway.

However, this is only relevant to MinVer when more than one tag is on the same commit, in which case MinVer will use the one with the latest version. So if there are two tags which only differ in case, the one that is chosen may not necessarily be the latest version according to the SemVer 2.0 spec. But this is such an unlikely scenario that I doubt anyone will ever be affected by it, nor will they probably even care which tag is chosen from those which differ only in case.

**Reply by teneko** [reply]
I sticked first with NuGet.Versioning too, but the problem was I had constantly to convert between my own SemanticVersion and NuGet's SematicVersion just to use their TryParse because of abstracting away Nuget.Versioning at all. The result was an ugly record wrapping somehow NuGet's SemanticVersion by lazy initializing it. In the end I have written the parser to parse the components of a SemVer version without any dependency to any SemanticVersion class. So you're right, I just wanted you let to know that there is something to parse that is not bound to any SemanticVersion class. :)

I would also rather to stick to NuGet's SemanticVersion when I would not write my own Semantic Versioning piece of software like MinVer or GitVersion. :)

Thanks by the way for this great piece of software, I mean MinVer I love the simple approach. My vision for my software that is highly inspired by MinVer is to have it a little more managment from the "outside". For example I mean caching or single-branch-scoped versioning conventions. I like the way you think having MinVer strictly scoped to MSBuild and make it usable over MSBuild properties, but I think, there is some kind of potencial left while not get too complicated like GitVersion.

**Reply by adamralph** [reply]
> I just wanted you let to know that there is something to parse that is not bound to any SemanticVersion class. :)

Bear in mind that `NuGet.Versioning` is an _internal_ dependency for MinVer (from 4.0.0 onwards), so using `NuGet.Versioning.SemanticVersion` directly doesn't bind consumers of the MinVer package to that type. At any time, I could swap out `NuGet.Versioning` for something else without affecting consumers of the MinVer package. And in fact, `NuGet.Versioning.SemanticVersion` is useful because it takes care of other things in the SemVer spec like comparisons (albeit with the minor off-spec behaviour mentioned above).

> Thanks by the way for this great piece of software, I mean MinVer I love the simple approach.

I'm glad you find it useful :slightly_smiling_face:

> I like the way you think having MinVer strictly scoped to MSBuild and make it usable over MSBuild properties

I'm not sure if you are already aware, but the [minver-cli](https://www.nuget.org/packages/minver-cli) package provides a .NET tool which may be used independently of project files and SDK targets.

**Reply by teneko** [reply]
> Bear in mind that NuGet.Versioning is an internal [..]

Oh that's totally fine. :)

> I'm not sure if you are already aware [..]

Yes I am aware. :) Thanks for the offer. I tried it out and it works not out of the box. After specifying the prefix (-t v) minver-cli finds my tags but unfortunately cuts of my prefix in the result, so it does not get preserved unless told being discarded.

Then when I would just stick to minver-cli I would like to do all MSBuild stuff manually. Specifying NuGet version, assembly version and so on. I would like to use github workflow, a shell script what else to represent some handy branch versioning strategies. My vision is to reduce that kind of boilerplate as much as possible while being as extensible as possible.

But nevertheless I appreciate what MinVer is at the current state. Please keep it up. :)

**Reply by adamralph** [reply]
> I tried it out and it works not out of the box. After specifying the prefix (-t v) minver-cli finds my tags but unfortunately cuts of my prefix in the result, so it does not get preserved unless told being discarded.

That is by design. `minver-cli` produces a SemVer 2.0 version. A tag prefix cannot be part of the version, since that would not be a valid SemVer 2.0 version.

**Reply by teneko** [reply]
That's ok. I think I am sticking to the word convention because at the end of the day a prefix found in a valid semantic version is just as it is a convention I can assume. But in terms of normalization you would like to discard the prefix or even replace it. So we both have a different understanding what a versioning tool should be capable of and that's ok! :)

**Reply by adamralph** [reply]
I don't understand, but never mind.
